M2E-core build fails with Uses-constraint-violations

The Maven build of m2e-core fails when the latest Tycho 2.5.0-SNAPSHOT is used because the resolution of the Target-Platform fails because of `org.apache.felix.resolver.reason.ReasonException: Uses constraint violations`.

A draft PR in m2e outlines this issue: https://github.com/eclipse-m2e/m2e-core/pull/334:

```
[INFO] Fetching org.w3c.dom.smil_1.0.1.v200903091627.jar from https://download.eclipse.org/eclipse/updates/4.21/R-4.21-202109060500/plugins/ (21.24kB)
[ERROR] Internal error: java.lang.RuntimeException: org.osgi.framework.BundleException: Bundle org.eclipse.m2e.core.ui cannot be resolved:org.eclipse.m2e.core.ui [115]
[ERROR]   Unresolved requirement: Require-Bundle: org.eclipse.m2e.model.edit; bundle-version="[1.16.0,2.0.0)"; visibility:="reexport"
[ERROR]     -> Bundle-SymbolicName: org.eclipse.m2e.model.edit; bundle-version="1.18.2.qualifier"; singleton:="true"
[ERROR]        org.eclipse.m2e.model.edit [116]
[ERROR]          Unresolved requirement: Require-Bundle: org.eclipse.wst.xml.core; visibility:="reexport"
[ERROR]            -> Bundle-SymbolicName: org.eclipse.wst.xml.core; bundle-version="1.2.300.v202102110535"; singleton:="true"
[ERROR]               org.eclipse.wst.xml.core [105]
[ERROR]                 Unresolved requirement: Require-Bundle: org.eclipse.jem.util; bundle-version="[2.1.200,3.0.0)"; resolution:="optional"
[ERROR]                 Unresolved requirement: Require-Bundle: org.eclipse.wst.validation; bundle-version="[1.2.700,1.3.0)"; resolution:="optional"
[ERROR]                 Unresolved requirement: Require-Bundle: org.eclipse.wst.common.emfworkbench.integration; bundle-version="[1.2.100,1.3.0)"; resolution:="optional"
[ERROR]                 Bundle was not resolved because of a uses constraint violation.
[ERROR]   org.apache.felix.resolver.reason.ReasonException: Uses constraint violation. Unable to resolve resource org.eclipse.wst.xml.core [osgi.identity; osgi.identity="org.eclipse.wst.xml.core"; type="osgi.bundle"; version:Version="1.2.300.v202102110535"; singleton:="true"] because it is exposed to package 'org.w3c.dom.events' from resources org.eclipse.osgi [osgi.identity; osgi.identity="org.eclipse.osgi"; type="osgi.bundle"; version:Version="3.17.0.v20210823-1805"; singleton:="true"] and org.w3c.dom.events [osgi.identity; osgi.identity="org.w3c.dom.events"; type="osgi.bundle"; version:Version="3.0.0.draft20060413_v201105210656"] via two dependency chains.
[ERROR] 
[ERROR] Chain 1:
[ERROR]   org.eclipse.wst.xml.core [osgi.identity; osgi.identity="org.eclipse.wst.xml.core"; type="osgi.bundle"; version:Version="1.2.300.v202102110535"; singleton:="true"]
[ERROR]     require: (&(osgi.wiring.bundle=org.eclipse.core.runtime)(&(bundle-version>=3.14.0)(!(bundle-version>=4.0.0))))
[ERROR]      |
[ERROR]     provide: osgi.wiring.bundle: org.eclipse.core.runtime
[ERROR]   org.eclipse.osgi [osgi.identity; osgi.identity="org.eclipse.osgi"; type="osgi.bundle"; version:Version="3.17.0.v20210823-1805"; singleton:="true"]
[ERROR] 
[ERROR] Chain 2:
[ERROR]   org.eclipse.wst.xml.core [osgi.identity; osgi.identity="org.eclipse.wst.xml.core"; type="osgi.bundle"; version:Version="1.2.300.v202102110535"; singleton:="true"]
[ERROR]     require: (&(osgi.wiring.bundle=org.apache.xerces)(&(bundle-version>=2.12.1)(!(bundle-version>=3.0.0))))
[ERROR]      |
[ERROR]     provide: osgi.wiring.bundle; bundle-version:Version="2.12.1.v20210115-0812"; osgi.wiring.bundle="org.apache.xerces"
[ERROR]   org.apache.xerces [osgi.identity; osgi.identity="org.apache.xerces"; type="osgi.bundle"; version:Version="2.12.1.v20210115-0812"]
[ERROR]     import: (osgi.wiring.package=org.w3c.dom.events)
[ERROR]      |
[ERROR]     export: osgi.wiring.package: org.w3c.dom.events
[ERROR]   org.w3c.dom.events [osgi.identity; osgi.identity="org.w3c.dom.events"; type="osgi.bundle"; version:Version="3.0.0.draft20060413_v201105210656"]
[ERROR] -> [Help 1]
org.apache.maven.InternalErrorException: Internal error: java.lang.RuntimeException: org.osgi.framework.BundleException: Bundle org.eclipse.m2e.core.ui cannot be resolved:org.eclipse.m2e.core.ui [115]
  Unresolved requirement: Require-Bundle: org.eclipse.m2e.model.edit; bundle-version="[1.16.0,2.0.0)"; visibility:="reexport"
    -> Bundle-SymbolicName: org.eclipse.m2e.model.edit; bundle-version="1.18.2.qualifier"; singleton:="true"
       org.eclipse.m2e.model.edit [116]
         Unresolved requirement: Require-Bundle: org.eclipse.wst.xml.core; visibility:="reexport"
           -> Bundle-SymbolicName: org.eclipse.wst.xml.core; bundle-version="1.2.300.v202102110535"; singleton:="true"
              org.eclipse.wst.xml.core [105]
                Unresolved requirement: Require-Bundle: org.eclipse.jem.util; bundle-version="[2.1.200,3.0.0)"; resolution:="optional"
                Unresolved requirement: Require-Bundle: org.eclipse.wst.validation; bundle-version="[1.2.700,1.3.0)"; resolution:="optional"
                Unresolved requirement: Require-Bundle: org.eclipse.wst.common.emfworkbench.integration; bundle-version="[1.2.100,1.3.0)"; resolution:="optional"
                Bundle was not resolved because of a uses constraint violation.
  org.apache.felix.resolver.reason.ReasonException: Uses constraint violation. Unable to resolve resource org.eclipse.wst.xml.core [osgi.identity; osgi.identity="org.eclipse.wst.xml.core"; type="osgi.bundle"; version:Version="1.2.300.v202102110535"; singleton:="true"] because it is exposed to package 'org.w3c.dom.events' from resources org.eclipse.osgi [osgi.identity; osgi.identity="org.eclipse.osgi"; type="osgi.bundle"; version:Version="3.17.0.v20210823-1805"; singleton:="true"] and org.w3c.dom.events [osgi.identity; osgi.identity="org.w3c.dom.events"; type="osgi.bundle"; version:Version="3.0.0.draft20060413_v201105210656"] via two dependency chains.

Chain 1:
  org.eclipse.wst.xml.core [osgi.identity; osgi.identity="org.eclipse.wst.xml.core"; type="osgi.bundle"; version:Version="1.2.300.v202102110535"; singleton:="true"]
    require: (&(osgi.wiring.bundle=org.eclipse.core.runtime)(&(bundle-version>=3.14.0)(!(bundle-version>=4.0.0))))
     |
    provide: osgi.wiring.bundle: org.eclipse.core.runtime
  org.eclipse.osgi [osgi.identity; osgi.identity="org.eclipse.osgi"; type="osgi.bundle"; version:Version="3.17.0.v20210823-1805"; singleton:="true"]

Chain 2:
  org.eclipse.wst.xml.core [osgi.identity; osgi.identity="org.eclipse.wst.xml.core"; type="osgi.bundle"; version:Version="1.2.300.v202102110535"; singleton:="true"]
    require: (&(osgi.wiring.bundle=org.apache.xerces)(&(bundle-version>=2.12.1)(!(bundle-version>=3.0.0))))
     |
    provide: osgi.wiring.bundle; bundle-version:Version="2.12.1.v20210115-0812"; osgi.wiring.bundle="org.apache.xerces"
  org.apache.xerces [osgi.identity; osgi.identity="org.apache.xerces"; type="osgi.bundle"; version:Version="2.12.1.v20210115-0812"]
    import: (osgi.wiring.package=org.w3c.dom.events)
     |
    export: osgi.wiring.package: org.w3c.dom.events
  org.w3c.dom.events [osgi.identity; osgi.identity="org.w3c.dom.events"; type="osgi.bundle"; version:Version="3.0.0.draft20060413_v201105210656"]
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:120)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.RuntimeException: org.osgi.framework.BundleException: Bundle org.eclipse.m2e.core.ui cannot be resolved:org.eclipse.m2e.core.ui [115]
  Unresolved requirement: Require-Bundle: org.eclipse.m2e.model.edit; bundle-version="[1.16.0,2.0.0)"; visibility:="reexport"
    -> Bundle-SymbolicName: org.eclipse.m2e.model.edit; bundle-version="1.18.2.qualifier"; singleton:="true"
       org.eclipse.m2e.model.edit [116]
         Unresolved requirement: Require-Bundle: org.eclipse.wst.xml.core; visibility:="reexport"
           -> Bundle-SymbolicName: org.eclipse.wst.xml.core; bundle-version="1.2.300.v202102110535"; singleton:="true"
              org.eclipse.wst.xml.core [105]
                Unresolved requirement: Require-Bundle: org.eclipse.jem.util; bundle-version="[2.1.200,3.0.0)"; resolution:="optional"
                Unresolved requirement: Require-Bundle: org.eclipse.wst.validation; bundle-version="[1.2.700,1.3.0)"; resolution:="optional"
                Unresolved requirement: Require-Bundle: org.eclipse.wst.common.emfworkbench.integration; bundle-version="[1.2.100,1.3.0)"; resolution:="optional"
                Bundle was not resolved because of a uses constraint violation.
  org.apache.felix.resolver.reason.ReasonException: Uses constraint violation. Unable to resolve resource org.eclipse.wst.xml.core [osgi.identity; osgi.identity="org.eclipse.wst.xml.core"; type="osgi.bundle"; version:Version="1.2.300.v202102110535"; singleton:="true"] because it is exposed to package 'org.w3c.dom.events' from resources org.eclipse.osgi [osgi.identity; osgi.identity="org.eclipse.osgi"; type="osgi.bundle"; version:Version="3.17.0.v20210823-1805"; singleton:="true"] and org.w3c.dom.events [osgi.identity; osgi.identity="org.w3c.dom.events"; type="osgi.bundle"; version:Version="3.0.0.draft20060413_v201105210656"] via two dependency chains.

Chain 1:
  org.eclipse.wst.xml.core [osgi.identity; osgi.identity="org.eclipse.wst.xml.core"; type="osgi.bundle"; version:Version="1.2.300.v202102110535"; singleton:="true"]
    require: (&(osgi.wiring.bundle=org.eclipse.core.runtime)(&(bundle-version>=3.14.0)(!(bundle-version>=4.0.0))))
     |
    provide: osgi.wiring.bundle: org.eclipse.core.runtime
  org.eclipse.osgi [osgi.identity; osgi.identity="org.eclipse.osgi"; type="osgi.bundle"; version:Version="3.17.0.v20210823-1805"; singleton:="true"]

Chain 2:
  org.eclipse.wst.xml.core [osgi.identity; osgi.identity="org.eclipse.wst.xml.core"; type="osgi.bundle"; version:Version="1.2.300.v202102110535"; singleton:="true"]
    require: (&(osgi.wiring.bundle=org.apache.xerces)(&(bundle-version>=2.12.1)(!(bundle-version>=3.0.0))))
     |
    provide: osgi.wiring.bundle; bundle-version:Version="2.12.1.v20210115-0812"; osgi.wiring.bundle="org.apache.xerces"
  org.apache.xerces [osgi.identity; osgi.identity="org.apache.xerces"; type="osgi.bundle"; version:Version="2.12.1.v20210115-0812"]
    import: (osgi.wiring.package=org.w3c.dom.events)
     |
    export: osgi.wiring.package: org.w3c.dom.events
  org.w3c.dom.events [osgi.identity; osgi.identity="org.w3c.dom.events"; type="osgi.bundle"; version:Version="3.0.0.draft20060413_v201105210656"]
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getResolverState (OsgiBundleProject.java:329)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.resolveClassPath (OsgiBundleProject.java:196)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.resolveProject (DefaultTychoResolver.java:173)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.lambda$resolveProjects$0 (TychoMavenLifecycleParticipant.java:127)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.accept (ForEachOps.java:183)
    at java.util.stream.WhileOps$1$1.accept (WhileOps.java:99)
    at java.util.ArrayList$ArrayListSpliterator.tryAdvance (ArrayList.java:1631)
    at java.util.stream.ReferencePipeline.forEachWithCancel (ReferencePipeline.java:127)
    at java.util.stream.AbstractPipeline.copyIntoWithCancel (AbstractPipeline.java:502)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:488)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:474)
    at java.util.stream.ForEachOps$ForEachOp.evaluateSequential (ForEachOps.java:150)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential (ForEachOps.java:173)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.forEach (ReferencePipeline.java:497)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.resolveProjects (TychoMavenLifecycleParticipant.java:158)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.afterProjectsRead (TychoMavenLifecycleParticipant.java:102)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:264)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.osgi.framework.BundleException: Bundle org.eclipse.m2e.core.ui cannot be resolved:org.eclipse.m2e.core.ui [115]
  Unresolved requirement: Require-Bundle: org.eclipse.m2e.model.edit; bundle-version="[1.16.0,2.0.0)"; visibility:="reexport"
    -> Bundle-SymbolicName: org.eclipse.m2e.model.edit; bundle-version="1.18.2.qualifier"; singleton:="true"
       org.eclipse.m2e.model.edit [116]
         Unresolved requirement: Require-Bundle: org.eclipse.wst.xml.core; visibility:="reexport"
           -> Bundle-SymbolicName: org.eclipse.wst.xml.core; bundle-version="1.2.300.v202102110535"; singleton:="true"
              org.eclipse.wst.xml.core [105]
                Unresolved requirement: Require-Bundle: org.eclipse.jem.util; bundle-version="[2.1.200,3.0.0)"; resolution:="optional"
                Unresolved requirement: Require-Bundle: org.eclipse.wst.validation; bundle-version="[1.2.700,1.3.0)"; resolution:="optional"
                Unresolved requirement: Require-Bundle: org.eclipse.wst.common.emfworkbench.integration; bundle-version="[1.2.100,1.3.0)"; resolution:="optional"
                Bundle was not resolved because of a uses constraint violation.
  org.apache.felix.resolver.reason.ReasonException: Uses constraint violation. Unable to resolve resource org.eclipse.wst.xml.core [osgi.identity; osgi.identity="org.eclipse.wst.xml.core"; type="osgi.bundle"; version:Version="1.2.300.v202102110535"; singleton:="true"] because it is exposed to package 'org.w3c.dom.events' from resources org.eclipse.osgi [osgi.identity; osgi.identity="org.eclipse.osgi"; type="osgi.bundle"; version:Version="3.17.0.v20210823-1805"; singleton:="true"] and org.w3c.dom.events [osgi.identity; osgi.identity="org.w3c.dom.events"; type="osgi.bundle"; version:Version="3.0.0.draft20060413_v201105210656"] via two dependency chains.

Chain 1:
  org.eclipse.wst.xml.core [osgi.identity; osgi.identity="org.eclipse.wst.xml.core"; type="osgi.bundle"; version:Version="1.2.300.v202102110535"; singleton:="true"]
    require: (&(osgi.wiring.bundle=org.eclipse.core.runtime)(&(bundle-version>=3.14.0)(!(bundle-version>=4.0.0))))
     |
    provide: osgi.wiring.bundle: org.eclipse.core.runtime
  org.eclipse.osgi [osgi.identity; osgi.identity="org.eclipse.osgi"; type="osgi.bundle"; version:Version="3.17.0.v20210823-1805"; singleton:="true"]

Chain 2:
  org.eclipse.wst.xml.core [osgi.identity; osgi.identity="org.eclipse.wst.xml.core"; type="osgi.bundle"; version:Version="1.2.300.v202102110535"; singleton:="true"]
    require: (&(osgi.wiring.bundle=org.apache.xerces)(&(bundle-version>=2.12.1)(!(bundle-version>=3.0.0))))
     |
    provide: osgi.wiring.bundle; bundle-version:Version="2.12.1.v20210115-0812"; osgi.wiring.bundle="org.apache.xerces"
  org.apache.xerces [osgi.identity; osgi.identity="org.apache.xerces"; type="osgi.bundle"; version:Version="2.12.1.v20210115-0812"]
    import: (osgi.wiring.package=org.w3c.dom.events)
     |
    export: osgi.wiring.package: org.w3c.dom.events
  org.w3c.dom.events [osgi.identity; osgi.identity="org.w3c.dom.events"; type="osgi.bundle"; version:Version="3.0.0.draft20060413_v201105210656"]
    at org.eclipse.tycho.core.osgitools.EquinoxResolver.assertResolved (EquinoxResolver.java:451)
    at org.eclipse.tycho.core.osgitools.EquinoxResolver.newResolvedState (EquinoxResolver.java:134)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.getResolverState (OsgiBundleProject.java:326)
    at org.eclipse.tycho.core.osgitools.OsgiBundleProject.resolveClassPath (OsgiBundleProject.java:196)
    at org.eclipse.tycho.core.resolver.DefaultTychoResolver.resolveProject (DefaultTychoResolver.java:173)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.lambda$resolveProjects$0 (TychoMavenLifecycleParticipant.java:127)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.accept (ForEachOps.java:183)
    at java.util.stream.WhileOps$1$1.accept (WhileOps.java:99)
    at java.util.ArrayList$ArrayListSpliterator.tryAdvance (ArrayList.java:1631)
    at java.util.stream.ReferencePipeline.forEachWithCancel (ReferencePipeline.java:127)
    at java.util.stream.AbstractPipeline.copyIntoWithCancel (AbstractPipeline.java:502)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:488)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:474)
    at java.util.stream.ForEachOps$ForEachOp.evaluateSequential (ForEachOps.java:150)
    at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential (ForEachOps.java:173)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:234)
    at java.util.stream.ReferencePipeline.forEach (ReferencePipeline.java:497)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.resolveProjects (TychoMavenLifecycleParticipant.java:158)
    at org.eclipse.tycho.core.maven.TychoMavenLifecycleParticipant.afterProjectsRead (TychoMavenLifecycleParticipant.java:102)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:264)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:957)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:289)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:193)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/InternalErrorException
```

--- a/tycho-core/src/main/java/org/eclipse/tycho/core/osgitools/EquinoxResolver.java
+++ b/tycho-core/src/main/java/org/eclipse/tycho/core/osgitools/EquinoxResolver.java
@@ -9,6 +9,7 @@
  *    Sonatype Inc. - initial API and implementation
  *    Christoph Läubrich -  [Bug 567782] Platform specific fragment not support in Multi-Platform POMless build
  *                          [Bug 572481] Tycho does not understand "additional.bundles" directive in build.properties
+ *                          [Issue 303] M2E-core build fails with Uses-constraint-violations
  *******************************************************************************/
 package org.eclipse.tycho.core.osgitools;
 
@@ -21,6 +22,9 @@ import java.util.List;
 import java.util.Map;
 import java.util.Properties;
 import java.util.StringJoiner;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
@@ -74,6 +78,25 @@ import org.osgi.framework.wiring.BundleRevision;
 
 @Component(role = EquinoxResolver.class)
 public class EquinoxResolver {
+
+    //The following properties are not supported for general usage and intended to experimental testing when developing Tycho, so use with care
+
+    /**
+     * Keep the default batch size, but allow to override this if necessary
+     */
+    private static final String BATCH_SIZE = System.getProperty("tycho.equinox.resolver.batch.size", null);
+    /**
+     * Set the batch timeout to an acceptable timeout before fallback to resolve one bundle at a
+     * time, but allow to override this if necessary
+     */
+    private static final String BATCH_TIMEOUT = System.getProperty("tycho.equinox.resolver.batch.timeout",
+            String.valueOf(TimeUnit.SECONDS.toMillis(30)));
+
+    /**
+     * Allow to adjust the default thread count used in resolver operations
+     */
+    private static final int THREAD_COUNT = Integer.getInteger("tycho.equinox.resolver.executor.threads", 1);
+
     @Requirement
     private BundleReader manifestReader;
 
@@ -88,64 +111,75 @@ public class EquinoxResolver {
 
     public ModuleContainer newResolvedState(ReactorProject project, MavenSession mavenSession, ExecutionEnvironment ee,
             DependencyArtifacts artifacts) throws BundleException {
-        Properties properties = getPlatformProperties(project, mavenSession, artifacts, ee);
-        ModuleContainer container = newState(artifacts, properties, mavenSession);
-        ResolutionReport report = container.resolve(null, false);
-        Module module = container.getModule(getNormalizedPath(project.getBasedir()));
-        if (module == null) {
-            Module systemModule = container.getModule(Constants.SYSTEM_BUNDLE_LOCATION);
-            if (project.getBasedir().equals(systemModule.getCurrentRevision().getRevisionInfo())) {
-                module = systemModule;
-            }
-        }
-        ModuleRevision moduleRevision = module.getCurrentRevision();
-        if (report.getEntries().isEmpty()) {
-            Collection<org.eclipse.osgi.container.Module> toUninstall = null;
-            if ((moduleRevision.getTypes() & BundleRevision.TYPE_FRAGMENT) != 0) {
-                // other fragments for the same host
-                toUninstall = moduleRevision.getWiring().getRequiredModuleWires(HostNamespace.HOST_NAMESPACE).stream()
-                        .map(ModuleWire::getProvider) //
-                        .flatMap(host -> host.getWiring().getProvidedModuleWires(HostNamespace.HOST_NAMESPACE).stream())
-                        .map(ModuleWire::getRequirer) //
-                        .filter(Predicate.not(moduleRevision::equals)) //
-                        .filter(fragment -> (fragment.getTypes() & BundleRevision.TYPE_FRAGMENT) != 0) //
-                        .map(revision -> revision.getRevisions().getModule()) //
-                        .collect(Collectors.toSet());
-            } else {
-                // fragments for the host
-                toUninstall = moduleRevision.getWiring().getProvidedModuleWires(HostNamespace.HOST_NAMESPACE).stream()
-                        .map(ModuleWire::getRequirer) //
-                        .filter(fragment -> (fragment.getTypes() & BundleRevision.TYPE_FRAGMENT) != 0) //
-                        .map(revision -> revision.getRevisions().getModule()) //
-                        .collect(Collectors.toSet());
-            }
-            if (!toUninstall.isEmpty()) {
-                for (Module uninstall : toUninstall) {
-                    try {
-                        container.uninstall(uninstall);
-                    } catch (BundleException e) {
-                        // TODO Auto-generated catch block
-                        e.printStackTrace();
-                    }
+        ScheduledExecutorService executorService = Executors.newScheduledThreadPool(THREAD_COUNT);
+        try {
+            Properties properties = getPlatformProperties(project, mavenSession, artifacts, ee);
+            ModuleContainer container = newState(artifacts, properties, mavenSession, executorService);
+            ResolutionReport report = container.resolve(null, false);
+            Module module = container.getModule(getNormalizedPath(project.getBasedir()));
+            if (module == null) {
+                Module systemModule = container.getModule(Constants.SYSTEM_BUNDLE_LOCATION);
+                if (project.getBasedir().equals(systemModule.getCurrentRevision().getRevisionInfo())) {
+                    module = systemModule;
                 }
-                report = container.refresh(null);
             }
-        }
-        assertResolved(report, moduleRevision);
+            ModuleRevision moduleRevision = module.getCurrentRevision();
+            if (report.getEntries().isEmpty()) {
+                Collection<org.eclipse.osgi.container.Module> toUninstall = null;
+                if ((moduleRevision.getTypes() & BundleRevision.TYPE_FRAGMENT) != 0) {
+                    // other fragments for the same host
+                    toUninstall = moduleRevision.getWiring().getRequiredModuleWires(HostNamespace.HOST_NAMESPACE)
+                            .stream().map(ModuleWire::getProvider) //
+                            .flatMap(host -> host.getWiring().getProvidedModuleWires(HostNamespace.HOST_NAMESPACE)
+                                    .stream())
+                            .map(ModuleWire::getRequirer) //
+                            .filter(Predicate.not(moduleRevision::equals)) //
+                            .filter(fragment -> (fragment.getTypes() & BundleRevision.TYPE_FRAGMENT) != 0) //
+                            .map(revision -> revision.getRevisions().getModule()) //
+                            .collect(Collectors.toSet());
+                } else {
+                    // fragments for the host
+                    toUninstall = moduleRevision.getWiring().getProvidedModuleWires(HostNamespace.HOST_NAMESPACE)
+                            .stream().map(ModuleWire::getRequirer) //
+                            .filter(fragment -> (fragment.getTypes() & BundleRevision.TYPE_FRAGMENT) != 0) //
+                            .map(revision -> revision.getRevisions().getModule()) //
+                            .collect(Collectors.toSet());
+                }
+                if (!toUninstall.isEmpty()) {
+                    for (Module uninstall : toUninstall) {
+                        try {
+                            container.uninstall(uninstall);
+                        } catch (BundleException e) {
+                            // TODO Auto-generated catch block
+                            e.printStackTrace();
+                        }
+                    }
+                    report = container.refresh(null);
+                }
+            }
+            assertResolved(report, moduleRevision);
 
-        return container;
+            return container;
+        } finally {
+            executorService.shutdownNow();
+        }
     }
 
     public ModuleContainer newResolvedState(File basedir, MavenSession mavenSession, ExecutionEnvironment ee,
             DependencyArtifacts artifacts) throws BundleException {
-        Properties properties = getPlatformProperties(new Properties(), mavenSession, null, ee);
-        ModuleContainer container = newState(artifacts, properties, mavenSession);
-        ResolutionReport report = container.resolve(null, false);
+        ScheduledExecutorService executorService = Executors.newScheduledThreadPool(THREAD_COUNT);
+        try {
+            Properties properties = getPlatformProperties(new Properties(), mavenSession, null, ee);
+            ModuleContainer container = newState(artifacts, properties, mavenSession, executorService);
+            ResolutionReport report = container.resolve(null, false);
 
-        ModuleRevision bundleDescription = container.getModule(getNormalizedPath(basedir)).getCurrentRevision();
-        assertResolved(report, bundleDescription);
+            ModuleRevision bundleDescription = container.getModule(getNormalizedPath(basedir)).getCurrentRevision();
+            assertResolved(report, bundleDescription);
 
-        return container;
+            return container;
+        } finally {
+            executorService.shutdownNow();
+        }
     }
 
     protected Properties getPlatformProperties(ReactorProject project, MavenSession mavenSession,
@@ -198,18 +232,19 @@ public class EquinoxResolver {
         return properties;
     }
 
-    protected ModuleContainer newState(DependencyArtifacts artifacts, Properties properties, MavenSession mavenSession)
-            throws BundleException {
+    protected ModuleContainer newState(DependencyArtifacts artifacts, Properties properties, MavenSession mavenSession,
+            ScheduledExecutorService executorService) throws BundleException {
         ModuleContainer[] moduleContainerAccessor = new ModuleContainer[1];
         ModuleContainerAdaptor moduleContainerAdaptor = new ModuleContainerAdaptor() {
 
             @Override
             public String getProperty(String key) {
+                // see https://github.com/eclipse/tycho/issues/213#issuecomment-912547700 for details about what these does
                 switch (key) {
-                // Prevent from too slow resolution, see
-                // https://github.com/eclipse/tycho/issues/213#issuecomment-912547700
                 case "equinox.resolver.revision.batch.size":
-                    return Integer.toString(1);
+                    return BATCH_SIZE;
+                case "equinox.resolver.batch.timeout":
+                    return BATCH_TIMEOUT;
                 }
                 return super.getProperty(key);
             }
@@ -286,6 +321,11 @@ public class EquinoxResolver {
                         // nothing to do
                     }
                 };
+            }
+
+            @Override
+            public ScheduledExecutorService getScheduledExecutor() {
+                return executorService;
             }
         };
         ModuleDatabase moduleDatabase = new ModuleDatabase(moduleContainerAdaptor);

--- a/tycho-core/src/test/java/org/eclipse/tycho/core/osgitools/EquinoxResolverTest.java
+++ b/tycho-core/src/test/java/org/eclipse/tycho/core/osgitools/EquinoxResolverTest.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2012, 2020 Sonatype Inc. and others.
+ * Copyright (c) 2012, 2021 Sonatype Inc. and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -57,7 +57,7 @@ public class EquinoxResolverTest extends AbstractTychoMojoTestCase {
 
     public void test_noSystemBundle() throws BundleException {
         Properties properties = subject.getPlatformProperties(new Properties(), null, null, DUMMY_EE);
-        ModuleContainer container = subject.newState(new DefaultDependencyArtifacts(), properties, null);
+        ModuleContainer container = subject.newState(new DefaultDependencyArtifacts(), properties, null, null);
         assertEquals(1, container.getModules().stream().map(Module::getCurrentRevision)
                 .map(BundleRevision::getSymbolicName).filter(Constants.SYSTEM_BUNDLE_SYMBOLICNAME::equals).count());
     }

